File: Project/MyProject/MyProject/Source/MyProject/Private/BasicFrame/Libs/MsgRoute/MsgRouteNotify.cpp
#include "MyProject.h"
#include "MsgRouteNotify.h"
#include "MsgRouteBase.h"
#include "MsgRouteDispatchHandle.h"

void MsgRouteNotify::addOneNotify(MsgRouteDispatchHandle* disp)
{
	if (this->mDispatchList.contains(disp))
	{
		this->mDispatchList.add(disp);
	}
}

void MsgRouteNotify::removeOneNotify(MsgRouteDispatchHandle* disp)
{
	if (this->mDispatchList.contains(disp))
	{
		this->mDispatchList.remove(disp);
	}
}

void MsgRouteNotify::handleMsg(MsgRouteBase* msg)
{
	for (auto item : this->mDispatchList.getList())
	{
		item->handleMsg(msg);
	}

	//Ctx.m_instance.mPoolSys.deleteObj(msg);
}

File: Project/MyProject/Tool/CSToolKit/Project/ToolKit/ToolKit/Source/Libs/MsgRoute/MsgRouteNotify.cs
namespace SDK.Lib
{
    public class MsgRouteNotify
    {
        protected MList<MsgRouteDispatchHandle> mDispatchList;

        public MsgRouteNotify()
        {
            this.mDispatchList = new MList<MsgRouteDispatchHandle>();
        }

        public void addOneNotify(MsgRouteDispatchHandle disp)
        {
            if(!this.mDispatchList.contains(disp))
            {
                this.mDispatchList.add(disp);
            }
        }

        public void removeOneNotify(MsgRouteDispatchHandle disp)
        {
            if(this.mDispatchList.contains(disp))
            {
                this.mDispatchList.remove(disp);
            }
        }

        public void handleMsg(MsgRouteBase msg)
        {
            int index = 0;
            int liseLen = this.mDispatchList.count();

            while(index < liseLen)
            {
                this.mDispatchList[index].handleMsg(msg);

                ++index;
            }

            // 暂时不用缓存，非资源数据结构重新申请内存应该不会太耗时
            //Ctx.msInstance.mPoolSys.deleteObj(msg);
        }
    }
}